src/input.rs

```diff
@@ -0,0 +1,116 @@
+use crate::model::*;
+use egg::*;
+use std::collections::HashMap;
+
+/// Struct for converting a model specified using our Rust interface to RecExpr
+///
+/// The RecExpr is growed on the fly when member functions are called. Uses a
+/// Hashmap to store the map of scalar nodes to their indices into the RexExpr to
+/// avoid replication.
+#[derive(Default)]
+pub struct GraphConverter {
+    rec_expr: RecExpr<Mdl>,
+    scalar_map: HashMap<i32, Id>,
+    name_gen: NameGen,
+}
+
+/// The APIs of GraphConverter are (intended to) match TASO's so that we can easily
+/// constructing TASO graphs using this class
+impl GraphConverter {
+    /// Gets the RexExpr after graph is constructed
+    pub fn rec_expr(self) -> RecExpr<Mdl> {
+        self.rec_expr
+    }
+
+    /// Takes in the parameters for the new input, construct the node in RexExpr,
+    /// return the Id (index) of this input node in the RecExpr. This is the
+    /// pattern for all these op functions.
+    pub fn new_input(&mut self, name: &str, dim1: i32, dim2: i32, dim3: i32, dim4: i32) -> Id {
```

Comment:
Why 4 dims? Would an array also work here, so that you could iterate over it?

Reply:
I'm currently specifying dimensions of input nodes in egraph with syntactic child nodes, so it won't directly support varying dimensions of inputs nodes. I can change to support variable dimension inputs later if some benchmarks need that. It's a quick implementation now for research purpose